File: add_two_numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// https://leetcode.com/problems/add-two-numbers/description/

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    res, sum := new(ListNode),0
    for cur := res;l1!=nil||l2!=nil||sum!=0;cur=cur.Next {
        if l1!=nil {
            sum += l1.Val
            l1=l1.Next
        }
        if l2!=nil {
            sum+=l2.Val
            l2=l2.Next
        }
        cur.Next=&ListNode{Val: (sum%10)}
        sum= sum/10
    }
    return res.Next
}

File: add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res=ListNode()
        sum=0
        cur=res
        while l1 is not None or l2 is not None or sum is not 0:
            if l1 is not None:
                sum=sum+l1.val
                l1=l1.next
            if l2 is not None:
                sum=sum+l2.val
                l2=l2.next
            cur.next=ListNode(sum%10)
            sum=sum//10
            cur=cur.next
        return res.next
